### ISSUE TITLE
Raised exception and inconsistent state of catalog when changing the type of an entity

### ISSUE THREAD

**Issue by jcarres-mdsol** [backstage/backstage#10321]
## Expected Behavior

The entity should not be there or at least allow me to update it manually.

## Actual Behavior

Blows up.
Stacktrace
Error: No path for routeRef{type=absolute,id=catalog}
    at useRouteRef (webpack-internal:///../../node_modules/@backstage/core-plugin-api/dist/index.esm.js:434:11)
    at EntityOrphanWarning (webpack-internal:///../../node_modules/@backstage/plugin-catalog/dist/esm/index-74617ebb.esm.js:930:94)
    at renderWithHooks (webpack-internal:///../../node_modules/@hot-loader/react-dom-v17/cjs/react-dom.development.js:14986:18)
    at mountIndeterminateComponent (webpack-internal:///../../node_modules/@hot-loader/react-dom-v17/cjs/react-dom.development.js:17812:13)
    at beginWork (webpack-internal:///../../node_modules/@hot-loader/react-dom-v17/cjs/react-dom.development.js:19050:16)
    at HTMLUnknownElement.callCallback (webpack-internal:///../../node_modules/@hot-loader/react-dom-v17/cjs/react-dom.development.js:3946:14)
    at Object.invokeGuardedCallbackDev (webpack-internal:///../../node_modules/@hot-loader/react-dom-v17/cjs/react-dom.development.js:3995:16)
    at invokeGuardedCallback (webpack-internal:///../../node_modules/@hot-loader/react-dom-v17/cjs/react-dom.development.js:4057:31)
    at beginWork$1 (webpack-internal:///../../node_modules/@hot-loader/react-dom-v17/cjs/react-dom.development.js:23960:7)
    at performUnitOfWork (webpack-internal:///../../node_modules/@hot-loader/react-dom-v17/cjs/react-dom.development.js:22775:12)

## Steps to Reproduce

1. Have a domain with name "mydomain" Domain in a github repo. Owned by "org" and with some systems referencing it.
2. Have a change of heart and make "mydomain" a Group instead of a Domain.
3. try to explore. "mydomain" will be still a domain, not a group, if clicking on it, the above stacktrace will appear.

## Context

I'm trying to go over the limitation of not being able to have more layers in the component/system/domain hierarchy by using groups instead.

## Your Environment

Data read as yaml files in github repositories.

- Output of `yarn backstage-cli info`: 

```text
OS:   Darwin 20.6.0 - darwin/x64
node: v16.14.0
yarn: 1.22.17
cli:  0.16.0 (installed)

Dependencies:
  @backstage/app-defaults                         0.2.1
  @backstage/backend-common                       0.13.0
  @backstage/backend-tasks                        0.2.0
  @backstage/catalog-client                       0.9.0
  @backstage/catalog-model                        0.13.0
  @backstage/cli-common                           0.1.8
  @backstage/cli                                  0.15.3
  @backstage/config-loader                        0.9.7
  @backstage/config                               0.1.15
  @backstage/core-app-api                         0.6.0
  @backstage/core-components                      0.9.1
  @backstage/core-plugin-api                      0.8.0
  @backstage/dev-utils                            0.2.25
  @backstage/errors                               0.2.2
  @backstage/integration-react                    0.1.25
  @backstage/integration                          0.8.0
  @backstage/plugin-api-docs                      0.8.2
  @backstage/plugin-app-backend                   0.3.29
  @backstage/plugin-auth-backend                  0.12.1
  @backstage/plugin-auth-node                     0.1.5
  @backstage/plugin-catalog-backend-module-aws    0.1.2
  @backstage/plugin-catalog-backend-module-github 0.1.0
  @backstage/plugin-catalog-backend               0.24.0
  @backstage/plugin-catalog-common                0.2.2
  @backstage/plugin-catalog-graph                 0.2.14
  @backstage/plugin-catalog-import                0.8.5
  @backstage/plugin-catalog-react                 0.9.0
  @backstage/plugin-catalog                       0.10.0
  @backstage/plugin-fossa                         0.2.34
  @backstage/plugin-github-actions                0.5.2
  @backstage/plugin-org                           0.5.2
  @backstage/plugin-permission-backend            0.5.4
  @backstage/plugin-permission-common             0.5.2
  @backstage/plugin-permission-node               0.5.4
  @backstage/plugin-permission-react              0.3.3
  @backstage/plugin-proxy-backend                 0.2.23
  @backstage/plugin-scaffolder-backend            0.18.0
  @backstage/plugin-scaffolder-common             0.3.0
  @backstage/plugin-scaffolder                    0.15.0
  @backstage/plugin-search-backend-node           0.5.1
  @backstage/plugin-search-backend                0.4.7
  @backstage/plugin-search-common                 0.3.1
  @backstage/plugin-search                        0.7.3
  @backstage/plugin-user-settings                 0.4.1
  @backstage/release-manifests                    0.0.2
  @backstage/search-common                        0.3.1
  @backstage/test-utils                           0.3.0
  @backstage/theme                                0.2.15
  @backstage/types                                0.1.3
  @backstage/version-bridge                       0.1.2
```

**Comment by jcarres-mdsol** [comment]
I guess it is clear by the stacktrace but it happens within ` <EntityOrphanWarning />` . I.e. If I remove it from the page, the page will display.
Also I think it still raises even if the entities have different names.

**Comment by freben** [comment]
Thanks for reporting! We'll look into why it crashes.

But regarding the kind change: Note that kind, namespace and name all form a unique key triplet. You can't just change the kind of an existing entity - you now have TWO entities; the one with the old kind (orphaned) and the one with the new kind.

Also slightly worried about the use of Group. That one is really strictly meant to be used for organizational entities like teams. What are these additional layers that you need?

**Comment by jcarres-mdsol** [comment]
best explained here:
https://github.com/backstage/backstage/issues/6269#issuecomment-1072487547

Company named X has many
Departments each has many
Solution Lines each has many
Products each built using one or more
Service

Right now my plan is:
Service -> Component
Product -> System
Solution line (if exists) -> Domain
Department -> Group
Company -> Group

Solution lines sometimes do not exist, a department simply have many products. But for big departments they will be further divided.

**Comment by anisjonischkeit** [comment]
For anyone that comes across this issue, I tried bumping/fixing versions/mismatches. I had just created my own catalog index page and didn't follow the doc on creating a new catalog index page. This ended up being the fix: 

<img width="1088" alt="Screen Shot 2022-09-19 at 8 20 30 pm" src="https://user-images.githubusercontent.com/22971353/190997313-a82ea030-6d88-4b06-9c0d-6374ce65aafd.png">

Basically, you still want the CatalogIndexPage from backstage to be specified (because I guess it sets up these `path for routeRef`s?) and have your custom index page be a child of that route Element.

See: https://backstage.io/docs/features/software-catalog/catalog-customization
